plugins/mongodb/mongodbembed.hpp

```diff
@@ -353,39 +383,60 @@ namespace mongodbembed
         }
 
         /**
-         * @brief Configures the MongoDB instance and sets up the MongoDB pool. It 
-         * should only be called once because only one instance object is allowed at
-         * a time.
+         * @brief Configures the MongoDB instance for the client objects to use for connections. It 
+         * should only be called once because only one instance object is allowed per program.
          * 
          * @param instance The instance object that is to be kept alive for multiple
          * threads to have access to.
-         * @param pool The pool object that has the information about which database
-         * and collection to connect to.
          */
-        void configure(std::unique_ptr<mongocxx::instance> && instance, std::unique_ptr<mongocxx::pool> && pool) 
+        void configure(std::unique_ptr<mongocxx::instance> && instance) 
         {
             _instance = std::move(instance);
-            _pool = std::move(pool);
         }
 
-        using connection = mongocxx::pool::entry;
+        /**
+         * @brief Creates a client object using the specified connections string. The client is
+         * added to the map of active connections.
+         * 
+         * @param connectionString The connection string for constructing the client object.
+         * 
+         * @param queryString A const char * holding the query string for hashing.
+         */
+        void create_connection(const char *connectionString, const char * queryString) 
+        {
+            std::shared_ptr<mongocxx::client> client_ptr = std::make_shared<mongocxx::client>(mongocxx::client{mongocxx::uri{connectionString}});
+
+            // Use a hash of the connection string as the key to finding
+            // any connection objects
+            std::string hashString = std::string(connectionString) + std::string(queryString);
+            hash64_t key = rtlHash64VStr(hashString.c_str(), 0); 
+
+            clientConnections[key] = client_ptr;
```

Comment:
Is this thread-safe?  I suspect reading and writing to clientConnections needs protecting with a critical section.

Reply:
Ok I did some digging, and the couchbase plugin definitely uses critical sections whenever it modifies the connection pool. I will get working on that.

Reply:
I don't think the critical section is quite correct.  There are two different questions

i) is the mongodb library thread safe when creating objects.
I assume that is true since most modern code takes this into account.

ii) Are the data structures in this file thread safe.
It is possible for multiple plugin functions to be called at the same time on different threads.  The clientConnections objects is potentially modified by 2 threads at the same time, or read at the same time that it is being written.

To protect against this you need a CriticalBlock around the code that reads from the map as well as the code that modifies it.  You should also minimize the length of time the critical lock is held (otherwise other threads can unnecessarily be held up).  If the mongodb library is thread safe then the criticalblock should be moved to just before the update of clientConnections.

Reply:
In the documentation they don't mention anything specific to creating client objects but from looking at the source code for the client constructor it doesn't seem to use any shared resources and creates everything from the uri passed in. I believe you are correct in that it the creation of the client is thread safe, but if you would like to look at the mongocxx code I can show you where that is. I added a CriticalBlock to where the client gets read from the map and removed the client creation from the other CriticalBlock.